source/sn_coap_protocol.c

```diff
@@ -1267,67 +1272,106 @@ static void sn_coap_protocol_linked_list_blockwise_payload_store(struct coap_s *
-        tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate payload!");
-        handle->sn_coap_protocol_free(stored_blockwise_payload_ptr);
-        return;
-    }
+        memcpy(stored_blockwise_payload_ptr->payload_ptr, temp_ptr, stored_blockwise_payload_ptr->payload_len);
+        memcpy(stored_blockwise_payload_ptr->payload_ptr + stored_blockwise_payload_ptr->payload_len, payload_ptr, payload_len);
+        stored_blockwise_payload_ptr->payload_len = new_len;
+        handle->sn_coap_protocol_free(temp_ptr);
 
-    /* Allocate memory for stored Payload's address */
-    stored_blockwise_payload_ptr->addr_ptr = sn_coap_protocol_malloc_copy(handle, addr_ptr->addr_ptr, addr_ptr->addr_len);
+    } else {
+        stored_blockwise_payload_ptr = NULL;
+        stored_blockwise_payload_ptr = handle->sn_coap_protocol_malloc(sizeof(coap_blockwise_payload_s));
 
-    if (stored_blockwise_payload_ptr->addr_ptr == NULL) {
-        tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate address pointer!");
-        handle->sn_coap_protocol_free(stored_blockwise_payload_ptr->payload_ptr);
-        handle->sn_coap_protocol_free(stored_blockwise_payload_ptr);
-        return;
-    }
+        if (stored_blockwise_payload_ptr == NULL) {
+            tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate blockwise!");
+            return;
+        }
 
-    /* Allocate & copy token number */
-    if (token_ptr && token_len) {
-        stored_blockwise_payload_ptr->token_ptr = sn_coap_protocol_malloc_copy(handle, token_ptr, token_len);
+        // If size1 option present then we can allocate the whole payload before hand.
+        stored_blockwise_payload_ptr->use_size1 = size1 ? 1 : 0;
 
-        if (!stored_blockwise_payload_ptr->token_ptr) {
-            tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate token pointer!");
-            handle->sn_coap_protocol_free(stored_blockwise_payload_ptr->addr_ptr);
+        /* Allocate memory for stored Payload's data */
+        if (stored_blockwise_payload_ptr->use_size1) {
+            stored_blockwise_payload_ptr->payload_ptr = handle->sn_coap_protocol_malloc(size1);
+            if (stored_blockwise_payload_ptr->payload_ptr) {
+                memcpy(stored_blockwise_payload_ptr->payload_ptr, payload_ptr, payload_len);
+                stored_blockwise_payload_ptr->payload_len = size1;
+            }
+        } else {
+            stored_blockwise_payload_ptr->payload_ptr = sn_coap_protocol_malloc_copy(handle, payload_ptr, payload_len);
+            stored_blockwise_payload_ptr->payload_len = payload_len;
+        }
+
+        if (stored_blockwise_payload_ptr->payload_ptr == NULL) {
+            tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate payload!");
+            stored_blockwise_payload_ptr->payload_len = 0;
```

Comment:
No need to clear struct just prior to freeing it.

Reply:
Fixed